Gems/PhysX/Code/Source/PhysXCharacters/Components/RagdollComponent.cpp

```diff
@@ -389,17 +406,62 @@ namespace PhysX
             }
         }
 
+        // If mass ratio clamping is enabled, iterate out from the root and clamp mass values
+        if (m_enableMassRatioClamping)
+        {
+            const float maxMassRatio = AZStd::GetMax(1.0f + AZ::Constants::FloatEpsilon, m_maxMassRatio);
+
+            // figure out the depth of each node in the tree, so that nodes can be visited from the root outwards
+            AZStd::vector<Utils::Characters::DepthData> nodeDepths =
+                Utils::Characters::ComputeHierarchyDepths(ragdollConfiguration.m_parentIndices);
+
+            AZStd::sort(
+                nodeDepths.begin(), nodeDepths.end(),
+                [](const Utils::Characters::DepthData& d1, const Utils::Characters::DepthData& d2)
+                {
+                    return d1.m_depth < d2.m_depth;
+                });
+
+            bool massesClamped = false;
+            for (const auto [depth, nodeIndex] : nodeDepths)
+            {
+                size_t parentIndex = ragdollConfiguration.m_parentIndices[nodeIndex];
+                if (parentIndex < numNodes)
+                {
+                    AzPhysics::RigidBody& nodeRigidBody = ragdoll->GetNode(nodeIndex)->GetRigidBody();
+                    const float originalMass = nodeRigidBody.GetMass();
+                    const float parentMass = ragdoll->GetNode(parentIndex)->GetRigidBody().GetMass();
+                    const float minMass = parentMass / maxMassRatio;
+                    const float maxMass = parentMass;
+                    if (originalMass < minMass || originalMass > maxMass)
+                    {
+                        const float clampedMass = AZStd::clamp(originalMass, minMass, maxMass);
+                        nodeRigidBody.SetMass(clampedMass);
+                        massesClamped = true;
+                        if (!AZ::IsCloseMag(originalMass, 0.0f))
+                        {
+                            // scale the inertia proportionally to how the mass was modified
+                            auto pxRigidBody = static_cast<physx::PxRigidDynamic*>(nodeRigidBody.GetNativePointer());
+                            pxRigidBody->setMassSpaceInertiaTensor(clampedMass / originalMass * pxRigidBody->getMassSpaceInertiaTensor());
```

Comment:
Just to double check if I understood what this is doing.
- If node's mass (originalMass) is larger than his parent's mass (maxPass) then the node's mass will be set to parent's mass and its inertia scaled down by a factor of (maxPass/originalMass, which is < 1)
- If node's mass (originalMass) is lower than his parent's mass by a ratio (minMass) then the node's mass will be set to that minMass and its inertia will be scaled up by a factor of (minPass/originalMass, which is > 1).

Is this correct?

Reply:
Yeah, that's right. It's just trying to scale the inertia the same way it does the mass, so they stay consistent when the mass is modified.